### PR TITLE
Quantum dot example

### PR DESCRIPTION
**Pull request by adamfrees**
Added example of electrode design as used in quantum-dot-based approaches to quantum computing. Design consists of three overlapping layers, each separated by a thin layer of oxide.

### CODE REVIEW

File: examples/quantum_dot_device/quantum_dot_device.py [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+
+# Parameters for geometry building
+input_file = "qd_device_parts.fcstd"
+input_parts = [substrate, gate1, gate2, gate3, gate4, wrap1, layer2, wrap2, layer3]
+
+# Compute parametrised geometries in parallel with dask
+geo = build_3d_geometry(input_parts=input_parts, input_file=input_file)
+
+# Create a local temporary directory to investigate results
+if not os.path.exists("tmp"):
```
Comment (basnijholt) [inline examples/quantum_dot_device/quantum_dot_device.py]: Instead of the if-statement you can do:
```
os.makedirs("tmp", exist_ok=True)
```

Comment (adamfrees) [reply examples/quantum_dot_device/quantum_dot_device.py]: Nice, I didn't know that! I've updated the code to get rid of the if-statement.

File: examples/quantum_dot_device/quantum_dot_device.py [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+# Compute parametrised geometries in parallel with dask
+geo = build_3d_geometry(input_parts=input_parts, input_file=input_file)
+
+# Create a local temporary directory to investigate results
+os.makedirs("tmp", exist_ok=True)
+
+print("Writing in directory tmp:")
+
+print("Writing parametrised instance to FreeCAD file.")
+geo.write_fcstd(os.path.join("tmp.fcstd"))
```
Comment (basnijholt) [inline examples/quantum_dot_device/quantum_dot_device.py]: `os.path.join("tmp.fcstd")` is not joining anything.

I guess you intended to do `os.path.join("tmp", "tmp.fcstd")`.

**Comment by likewei92** [comment]
Can this be checked in as a test (even if there're no assertions) instead of an example? Examples tend to get out of date (case in point: I have a refactor that's merging soon that'll change the geometry construction interfaces) whereas people are forced to go fix all the tests

**Comment by adamfrees** [comment]
@likewei92 This example takes about 15 minutes to run on my laptop - do you think that this will be too long to run as a test?

If so, maybe I could make a simpler version of this and make it a test in addition to the example as-is. I think having this particular example will be very instructive to people in my sub-field, as this would be a very common use-case.

**Comment by basnijholt** [comment]
I would say that's definitely too long, a test should just make sure that something runs correctly.

**Comment by likewei92** [comment]
Yeah that's definitely too long for a test. We probably want to think about how to maintain our examples so they don't get out of date though

**Comment by adamfrees** [comment]
I agree completely - one (admittedly not great) fix would be to add a comment pointing to the relevant test (test_geo_task). This example is pretty close to the "test_geo_task" test; the biggest difference is in the underlying FreeCAD objects.

File: examples/quantum_dot_device/quantum_dot_device.py (hunk 2 above)
Comment (adamfrees) [reply examples/quantum_dot_device/quantum_dot_device.py]: That's correct, thanks!

**Comment by johnkgamble** [comment]
Thanks, @adamfrees! Perhaps you could leave the example as-is, but also add a few tests to cover things. For example, you include the litho on only a few of the gates instead of all of them. That way, the example is still plausibly covered by the test. The example should be noted in these tests so we know it will need to be updated if we break the interface. Spoiler alert: breaking changes for the interface coming soon! :)

**Comment by likewei92** [comment]
Note that there have been some breaking changes to the geometry interface. Please merge from master and update to the new interface. Let me know if you need any help :)

**Comment by adamfrees** [comment]
@likewei92 @johnkgamble I've updated the example and added a simplified test that takes ~10 seconds, but for some reason the test fails to find the correct FCStd file in the automated testing (https://dev.azure.com/ms-quantum-public/9af4e09e-a436-4aca-9559-2094cfe8d80c/_build/results?buildId=1470) (pytest works just fine locally). Has this issue been seen before?

**Comment by adamfrees** [comment]
@likewei92 @johnkgamble Solved! Turns out the file had the suffix ".FCStd," not ".fcstd" - everything should be working properly now.

**Pull request merged**